### DISCUSSION TITLE
Ant Design Vue support

### DISCUSSION THREAD

**Discussion by fleximus** [nuxt/nuxt#16363] [General]
Looking forward to see [ant-design-vue](https://github.com/vueComponent/ant-design-vue) working with Nuxt 3 in the near future.

Antdv is hard working on v3.0 in Q4 2021, so I've asked in the roadmap thread to support Nuxt 3 - [please consider upvoting](https://github.com/vueComponent/ant-design-vue/discussions/4575#discussioncomment-1486487).

Also, it would be nice to see ant-design-vue listed in the [Module Support Status](https://github.com/nuxt/framework/discussions/751) list.

**Comment by wobsoriano** [comment]
Any problems if you install it manually?

```ts
// plugins/ant-design.ts
import Antd from 'ant-design-vue';

export default defineNuxtPlugin(nuxtApp => {
  nuxtApp.vueApp.use(Antd)
})
```

```ts
// nuxt.config.ts
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
    css: ['ant-design-vue/dist/antd.css'],
})
```

Working on my end.

**Reply by kakajansh** [reply]
As noted by @productdevbook `nuxtApp.app` was renamed to `nuxtApp.vueApp` #968

**Comment by fleximus** [comment]
Running still into this problem, not sure if anything is missing/needed?

```
SyntaxError: Cannot use import statement outside a module
    at wrapSafe (internal/modules/cjs/loader.js:1001:16)
    at Module._compile (internal/modules/cjs/loader.js:1049:27)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at ModuleWrap.<anonymous> (internal/modules/esm/translators.js:199:29)
    at ModuleJob.run (internal/modules/esm/module_job.js:183:25)
    at async Loader.import (internal/modules/esm/loader.js:178:24)
    at async __instantiateModule__ (file:///tmp/nuxt3_antdv/.nuxt/dist/server/server.mjs:2473:3)
Server Side Rendering Error: /tmp/nuxt3_antdv/node_modules/ant-design-vue/es/index.js:1
import * as components from './components';
```

**Reply by wobsoriano** [reply]
Try any of this:

```ts
import { defineNuxtConfig } from 'nuxt3';

export default defineNuxtConfig({
  vite: false
});
```

```ts
import { defineNuxtConfig } from 'nuxt3';

export default defineNuxtConfig({
  build: {
    transpile: ['ant-design-vue']
  }
});
```

**Comment by zndle** [comment]
try this:

```
// file : plugins/antd.ts
import {defineNuxtPlugin} from "#app";
import Antd from 'ant-design-vue/lib';

export default defineNuxtPlugin(nuxtApp => {
    nuxtApp.vueApp.use(Antd);
})
```

but some components doesn't work with `ssr` , for example : `a-menu` .

**Reply by ndrshn** [reply]
> try this:
> 
> ```
> // file : plugins/antd.ts
> import {defineNuxtPlugin} from "#app";
> import Antd from 'ant-design-vue/lib';
> 
> export default defineNuxtPlugin(nuxtApp => {
>     nuxtApp.vueApp.use(Antd);
> })
> ```
> 
> but some components doesn't work with `ssr` , for example : `a-menu` .

solved my issue. thanks :)

**Reply by li9269391** [reply]
> try this:
> 
> ```
> // file : plugins/antd.ts
> import {defineNuxtPlugin} from "#app";
> import Antd from 'ant-design-vue/lib';
> 
> export default defineNuxtPlugin(nuxtApp => {
>     nuxtApp.vueApp.use(Antd);
> })
> ```
> 
> but some components doesn't work with `ssr` , for example : `a-menu` .

run dev&build is OK, but run start will go wrong

**Comment by liziwork** [comment]
Can't use it 😫

**Comment by wobsoriano** [comment]
Updated working example:

1. Install deps

```
yarn add ant-design-vue@3.0.0-beta.8 && yarn add unplugin-vue-components -D
```

2. Update your `nuxt.config.ts` to this

```ts
import { defineNuxtConfig } from 'nuxt3';
import Components from 'unplugin-vue-components/vite';
import { AntDesignVueResolver } from 'unplugin-vue-components/resolvers';

export default defineNuxtConfig({
  vite: {
    plugins: [
      Components({
        resolvers: [AntDesignVueResolver()],
      }),
    ],
    // @ts-expect-error: Missing ssr key
    ssr: {
      noExternal: ['moment', 'compute-scroll-into-view', 'ant-design-vue'],
    },
  },
});
```

**Reply by Dekhnar** [reply]
Any news about this ?

**Reply by lvreninfo** [reply]
Any news about this ?

**Reply by wobsoriano** [reply]
This should work now. Updated my comment.

**Reply by vantd158** [reply]
Any news about this ?

**Reply by wobsoriano** [reply]
> Any news about this ?

The answer should be working

**Reply by xy406043** [reply]
take this way, it triggers a memory issue  when  nuxi build.
![error](https://yuxifile.watsonsestore.com.cn/group/14826752aead41dd9e406516e26df989.png)

**Comment by cenxky** [comment]
Here is another way I researched on nuxt3 doc.

1. Install ant-design-vue
```
yarn add ant-design-vue
```

2. Create `antdv.js` in `/modules` folder
```js
// modules/antdv.js
import { fileURLToPath } from "node:url"
import { defineNuxtModule } from "@nuxt/kit"

export default defineNuxtModule({
  setup (_, nuxt) {
    nuxt.options.css.push("ant-design-vue/dist/antd.css")
  },
  hooks: {
    "components:dirs" (dirs) {
      dirs.push({
        path: fileURLToPath(new URL("../node_modules/ant-design-vue/lib", import.meta.url)),
        pattern: "*/index.js",
        prefix: "a"
      })
    }
  }
})
```

3. Update Nuxt config
```ts
// nuxt.config.ts

export default defineNuxtConfig({
  ...
  modules: ["./modules/antdv"],
  ...
})
```
4. Use antd components in `app.vue` file
```vue
<template>
  <a-button>Here we are<a-button>
</template>
```

**Reply by suktec** [reply]
console return this
```
client.ts:28 SyntaxError: The requested module '/_nuxt/node_modules/ant-design-vue/lib/button/index.js' does not provide an export named 'default'
```

**Reply by u007** [reply]
i got this error:
```
Unexpected token 'export'

at Object.compileFunction (node:vm:352:18)
at wrapSafe (node:internal/modules/cjs/loader:1032:15)
at Module._compile (node:internal/modules/cjs/loader:1067:27)
at Object.Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
at Module.load (node:internal/modules/cjs/loader:981:32)
at Function.Module._load (node:internal/modules/cjs/loader:822:12)
at ModuleWrap. (node:internal/modules/esm/translators:168:29)
at ModuleJob.run (node:internal/modules/esm/module_job:197:25)
at processTicksAndRejections (node:internal/process/task_queues:96:5)
at async Promise.all (index 0)
```

**Comment by u007** [comment]
i got this working 
https://github.com/u007/nuxt-app

**Reply by Vostryakoff** [reply]
ant styles not work

**Reply by u007** [reply]
not all is working, some component will throw unknown error, very difficult to debug what causes it. turn out some component cannot auto import

**Reply by Vostryakoff** [reply]
<img width="519" alt="image" src="https://user-images.githubusercontent.com/27483492/175567442-b97e7395-74d9-4ab1-9ed2-73f31e0c153d.png">
may be is a bug on my ios mac. example a-button have not styles

**Reply by u007** [reply]
not sure man, i gave up using nuxt with ant.
u can manually import it
```
import 'ant-design-vue/es/button/style'
```

this is how i did it with vite

```
// inside vite.config.ts
import { createStyleImportPlugin, AntdResolve } from 'vite-plugin-style-import'

export default defineConfig({
  // other config
  ,plugins: [
    createStyleImportPlugin({
      resolves: [AntdResolve()],
      libs: [
        // If you don’t have the resolve you need, you can write it directly in the lib, or you can provide us with PR
        {
          libraryName: 'ant-design-vue',
          esModule: true,
          resolveStyle: (name) => {
            if (['notification', 'tab-pane', 'list-item'].indexOf(name) >= 0) {
              return ''
            }
            const mappedName = {
              'input-search': { path: 'input', name: 'search-input' },
            }
            const pathName = mappedName[name]?.path || name
            const fileName = mappedName[name]?.name || 'index.less'
            const jsFileName = 'index.js'
            const path = `ant-design-vue/es/${pathName}/style/${fileName}`
            const jsPath = `ant-design-vue/es/${pathName}/style/${jsFileName}`
            let exists = fs.existsSync('node_modules/'+path) || fs.existsSync('node_modules/'+jsPath)
            // console.log('importcss', name, path, exists, jsPath)
            if (fs.existsSync('node_modules/'+path)) {
              //file exists
              return path
            }
            if (fs.existsSync('node_modules/'+jsPath)) {
              return jsPath
            }
            console.log('importcss missing', name, path, jsPath)
            return ''
          },
        },
      ],
    }),
})
```

**Reply by Vostryakoff** [reply]
thx! but i'll wait for a stable release ant + nuxt3. In my case, the transfer of the project to nuxt3 is not possible without ant...

**Comment by kasvith** [comment]
It would be good to have some official support :(

**Comment by denisakp** [comment]
Hello guys, i dont know if this could help but i want to share with you the way i'm making my nuxt3 project work with ant design

### Step 1: Install Ant degin and ant design icons for vue

```sh
yarn add ant-design-vue @ant-design/icons-vue
```

### Step 2: Create a new file named antd.ts under the plugins folder (must be created if !exists) 
Copy and paste this code inside

```ts
import {defineNuxtPlugin} from "#app";
import 'ant-design-vue/dist/antd.css';
import Antd from 'ant-design-vue';

export default defineNuxtPlugin((nuxtApp) => {
    nuxtApp.vueApp.use(Antd)
})

```

### Step 3: Register your newly created plugin
To do so, open your nuxt.config.ts file and paste this code

``` ts
import { defineNuxtConfig } from 'nuxt';

// https://v3.nuxtjs.org/api/configuration/nuxt.config
export default defineNuxtConfig({
    plugins:['@/plugins/antd']
})
```
### Step 4: Enjoy
From now you can enjoy Ant design  by running your nuxtjs development server: 
```sh 
yarn dev
 ```
 POV: The current plugin register all components. If you need to register only fews or want to do any customization please refer to Ant design docs and update your plugin in consequence.

**Reply by kasvith** [reply]
but does this adds typings?

**Reply by denisakp** [reply]
@kasvith yes it does !

**Reply by kasvith** [reply]
let me see...if it does this is a great news

**Reply by gcentervn** [reply]
still working ???

**Reply by axelbdt** [reply]
This helped, thanks.
The `antd.css` file was renamed to `reset.css`.

Works with the following code:

```typescript
// plugins/antd.ts
import {defineNuxtPlugin} from "#app";
import 'ant-design-vue/dist/reset.css';
import Antd from 'ant-design-vue';

export default defineNuxtPlugin((nuxtApp) => {
    nuxtApp.vueApp.use(Antd)
})
```

Curious is there is a different recommended approach to Ant in Nuxt in 2024.

**Reply by TgkRuobin** [reply]
nice! it works

**Comment by arshx86** [comment]
unfortunately no method is works, the last one with plugins is working but causing too much memory leak and there is no types.

there is [a community module](https://nuxt.com/modules/ant-design-vue) that gets it into working with types but it has too much issues like memory leak and unsolvable hydration errors.

I dont recommend using antdesign until official module for nuxt pulls up :(
The most similar libraries is element plus and primevue.
